src/core/common/settings.cpp

```diff
@@ -0,0 +1,508 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the OpenThread platform abstraction for non-volatile storage of settings.
+ *
+ */
+
+#include <stdlib.h>
+#include <stddef.h>
+#include <string.h>
+#include <assert.h>
+
+#include <openthread-types.h>
+#include <openthread-core-config.h>
+
+#include <common/code_utils.hpp>
+#include <platform/settings.h>
+#include <platform/flash.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+enum
+{
+    kBlockAddFlag = 0x1,
+    kBlockDeleteFlag = 0x02,
+};
+
+enum
+{
+    kMaxStageAddNum = 3,
+    kMaxStageDeleteNum = 3,
+    kMaxStageDataLen = 32,
+};
+
+enum
+{
+    kSettingsFlagSize = 4,
+    kSettingsInSwap = 0xbe5cc5ef,
+    kSettingsInUse = 0xbe5cc5ee,
+    kSettingsNotUse = 0xbe5cc5ec,
+};
+
+OT_TOOL_PACKED_BEGIN
+struct settingsBlock
+{
+    uint16_t key;
+    uint8_t index;
+    uint8_t flag;
+    uint16_t length;
+    uint16_t reserved;
+} OT_TOOL_PACKED_END;
+
+OT_TOOL_PACKED_BEGIN
+struct stageAddSettingsBlock
+{
+    struct settingsBlock block;
+    uint8_t data[kMaxStageDataLen];
+} OT_TOOL_PACKED_END;
+
+OT_TOOL_PACKED_BEGIN
+struct stageDeleteSettingsBlock
+{
+    uint16_t key;
+    int index;
+} OT_TOOL_PACKED_END;
+
+static uint32_t sBase;
+static uint32_t sSettingsSize;
+static uint32_t sUsedSize;
+static bool sCommitLock = false;
+static bool sSwapDone = false;
+
+static uint16_t sStageActionSeq;
+static struct stageAddSettingsBlock sStageAddSettingsBlock[kMaxStageAddNum];
+static uint8_t sStageAddSettingsNum;
+static uint16_t sStageAddSettingsBufLength;
+static struct stageDeleteSettingsBlock sStageDeleteSettingsBlock[kMaxStageDeleteNum];
+static uint8_t sStageDeleteSettingsNum;
```

Comment:
I still keep using sBase and sUsedSize as global variables,and failed to find a way not using them as global...